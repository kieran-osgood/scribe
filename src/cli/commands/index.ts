// export * from './base-command';
// export * from './default-command';
export * from './init-command.js';
