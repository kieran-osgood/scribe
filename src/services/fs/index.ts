// https://github.com/Effect-TS/node/pull/629/commits/dbbfc3b0f6f7e0b21bcca5bd095612ea998f49b6
export * from './error.js';
export * from './fs.js';
