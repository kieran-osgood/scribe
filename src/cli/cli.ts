import { Command } from '@effect/cli';

import packageJson from '../../package.json';
import { ScribeInit } from './commands';
import { ScribeDefault } from './commands/default-command';

const command = ScribeDefault.pipe(Command.withSubcommands([ScribeInit]));

export const run = Command.run(command, {
  name: 'Scribe',
  version: packageJson.version,
});
