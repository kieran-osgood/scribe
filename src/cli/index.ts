import { Effect } from '@scribe/core';
import { Builtins, Cli } from 'clipanion';
import { BaseContext } from 'clipanion/lib/advanced/Cli';
import { DefaultCommand } from './commands/DefaultCommand';

export function _Cli(args: string[], contextOverrides?: Partial<BaseContext>) {
  const cli = new Cli({
    binaryLabel: `scribe`,
    binaryName: 'scribe',
    enableCapture: true,
  });

  cli.register(DefaultCommand);
  // cli.register(InitCommand);

  cli.register(Builtins.HelpCommand);
  cli.register(Builtins.VersionCommand);
  return cli.runExit(args, { ...Cli.defaultContext, ...contextOverrides });
}

export function run(args: string[], contextOverrides: Partial<BaseContext>) {
  return Effect.tryPromise(() => _Cli(args.slice(2), contextOverrides));
}
