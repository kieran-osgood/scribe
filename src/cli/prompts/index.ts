import { Prompt } from '@effect/cli';
import { Effect } from 'effect';

export const fileName = Prompt.text({
  message: 'Name:',
  validate: s =>
    /^([A-Za-z\-_\d])+$/.test(s)
      ? Effect.succeed(s)
      : Effect.fail(
          'File name may only include letters, numbers & underscores.',
        ),
});

export const continueWarning = Prompt.toggle({
  message: 'Continue?',
  active: 'yes',
  inactive: 'no',
});

export const templates = (s: string[]) =>
  Prompt.select({
    message: 'Template:',
    choices: s.map(_ => ({
      title: _,
      value: _,
      // https://github.com/kieran-osgood/scribe/issues/32
      // description: 'This is some description of a template',
    })),
  });
