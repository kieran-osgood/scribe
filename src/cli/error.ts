import { TaggedClass } from 'src/core';

export class CliError extends TaggedClass('CliError')<{
  readonly cause?: unknown;
}> {}
