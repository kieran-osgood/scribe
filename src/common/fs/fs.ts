import { Effect } from '@scribe/core';
import * as NFS from 'fs';
import { ErrnoError } from './error';

export const writeFile = (
  path: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
) =>
  Effect.async<never, ErrnoError, boolean>(resume => {
    NFS.writeFile(path, data, options, err => {
      if (err) {
        resume(Effect.fail(new ErrnoError(err)));
      } else {
        resume(Effect.succeed(true));
      }
    });
  });
