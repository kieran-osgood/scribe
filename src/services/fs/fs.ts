import { Abortable } from 'node:events';

import { Context, Effect, pipe } from 'effect';
import * as NFS from 'fs';
import * as memfs from 'memfs';
import path from 'path';

import { MkDirError, ReadFileError, StatError, WriteFileError } from './error';

export interface FS {
  writeFile: typeof NFS.writeFile;
  readFile: typeof NFS.readFile;
  mkdir: typeof NFS.mkdir;
  stat: typeof NFS.stat;
  open: typeof NFS.open;
  close: typeof NFS.close;
}

export const FS = Context.Tag<FS>();
export const FSLive = NFS;
export const FSMock = memfs.fs as unknown as typeof NFS;

export const getFS = (test: boolean) => {
  return test ? FSMock : FSLive;
};

export const writeFile = (
  file: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, WriteFileError, NFS.PathOrFileDescriptor>(resume => {
        fs.writeFile(file, data, options, error => {
          if (error) {
            resume(
              Effect.fail(new WriteFileError({ file, data, options, error })),
            );
          } else {
            resume(Effect.succeed(file));
          }
        });
      }),
    ),
  );

export const writeFileWithDir = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
): Effect.Effect<FS, WriteFileError | MkDirError, NFS.PathOrFileDescriptor> =>
  pipe(
    mkdir(path.dirname(pathName), { recursive: true }),
    Effect.flatMap(() => writeFile(pathName, data, options)),
  );

export const readFile = (
  path: NFS.PathOrFileDescriptor,
  options:
    | ({ encoding?: BufferEncoding; flag?: string } & Abortable)
    | undefined
    | null,
): Effect.Effect<FS, ReadFileError, string | Buffer> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, ReadFileError, string | Buffer>(resume => {
        fs.readFile(path, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new ReadFileError({ path, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const mkdir = (
  file: NFS.PathLike,
  options: NFS.MakeDirectoryOptions & {
    recursive: true;
  } = {
    // TODO!!!! Check mode
    mode: 0,
    recursive: true,
  },
): Effect.Effect<FS, MkDirError, string | undefined> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, MkDirError, string | undefined>(resume => {
        fs.mkdir(file, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new MkDirError({ file, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const stat = (path: string) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, StatError, NFS.Stats>(resume =>
        { fs.stat(path, (error, stats) => {
          if (error) {
            resume(Effect.fail(new StatError({ path, error })));
          } else {
            resume(Effect.succeed(stats));
          }
        }); },
      ),
    ),
  );

export const isFileOrDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile() || _.isDirectory()),
  );

export const isFile = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile()),
  );

export const isDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isDirectory()),
  );

// export const open = (
//   path: PathLike,
//   flags: OpenMode | undefined,
//   mode: Mode | undefined | null
// ): Effect.Effect<FS, OpenError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, OpenError, number>(resume =>
//         fs.open(path, flags, mode, (error, fd) => {
//           if (error) {
//             resume(Effect.fail(new OpenError({ path, flags, mode, error })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );

// export const close = (fd: number): Effect.Effect<FS, CloseError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, CloseError, number>(resume =>
//         fs.close(fd, error => {
//           if (error) {
//             resume(Effect.fail(new CloseError({ error, fd })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );
