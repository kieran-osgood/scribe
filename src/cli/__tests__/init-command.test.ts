import { Effect, Fiber, ReadonlyArray } from 'effect';
import fs from 'fs';
import path from 'path';

import packageJson from '../../../package.json';
import * as Cli from '../cli.js';
import { createMinimalProject, runEffect } from './fixtures.js';
import * as MockConsole from './mock-console.js';
import * as MockTerminal from './mock-terminal.js';

describe('[Given] Git Clean', () => {
  it('[When] user accepts [Then] create file', async () => {
    const cwd = createMinimalProject({
      git: { dirty: false, init: true },
      fixtures: { configFile: false, templateFiles: false, base: true },
    });

    return Effect.gen(function* ($) {
      const args = ReadonlyArray.make('init');
      const fiber = yield* $(Effect.fork(Cli.run(args)));

      yield* $(MockTerminal.inputKey('left'));
      yield* $(MockTerminal.inputKey('enter'));

      yield* $(Fiber.join(fiber));

      const lines = yield* $(MockConsole.getLines({ stripAnsi: true }));
      expect(lines).toMatchInlineSnapshot(`
          [
            "Init",
            " Git ",
            "Checking working tree clean",
            " Config ",
            "Checking write path clear",
            "Writing...",
            " Success ",
            "✅  Scribe init complete. Edit the config to begin templating.",
            "📁 file://${cwd}/scribe.config.ts",
          ]
        `);
      const config = fs.readFileSync(path.join(cwd, `scribe.config.ts`));
      expect(String(config)).toMatchSnapshot();
    }).pipe(runEffect(cwd));
  });

  it('[When] filepath full [then] print failure', async () => {
    const cwd = createMinimalProject();

    return Effect.gen(function* ($) {
      const config = fs.readFileSync(`${cwd}/scribe.config.ts`);
      const args = ReadonlyArray.make('init');
      const fiber = yield* $(Effect.fork(Cli.run(args)));

      yield* $(MockTerminal.inputKey('left'));
      yield* $(MockTerminal.inputKey('enter'));

      yield* $(Fiber.join(fiber));

      const lines = yield* $(MockConsole.getLines({ stripAnsi: true }));
      expect(lines).toMatchInlineSnapshot(`
          [
            "Init",
            " Git ",
            "Checking working tree clean",
            " Config ",
            "Checking write path clear",
            "Failed to create config. Path not empty.",
            "📁 file://${cwd}/scribe.config.ts",
          ]
        `);
      expect(String(config)).toMatchSnapshot();
    }).pipe(runEffect(cwd));
  });
});

describe('[Given] Git dirty', () => {
  describe('[Then] prompt user to continue', () => {
    it('[When] user accepts [Then] create file', async () => {
      const cwd = createMinimalProject({
        git: { dirty: true, init: true },
        fixtures: { configFile: false, templateFiles: false, base: true },
      });

      return Effect.gen(function* ($) {
        const args = ReadonlyArray.make('init');
        const fiber = yield* $(Effect.fork(Cli.run(args)));

        yield* $(MockTerminal.inputKey('left'));
        yield* $(MockTerminal.inputKey('enter'));

        yield* $(Fiber.join(fiber));

        const lines = yield* $(MockConsole.getLines({ stripAnsi: true }));
        expect(lines).toMatchInlineSnapshot(`
      [
        "Init",
        " Git ",
        "Checking working tree clean",
        "Git working tree dirty - proceed with caution.
      Recommendation: commit all changes before proceeding.",
        "? Continue? › yes / no",
        "? Continue? › yes / no",
        "✔ Continue? … yes / no
      ",
        "",
        " Config ",
        "Checking write path clear",
        "Writing...",
        " Success ",
        "✅  Scribe init complete. Edit the config to begin templating.",
        "📁 file://${cwd}/scribe.config.ts",
      ]
    `);
      }).pipe(runEffect(cwd));
    });

    it('[When] user declines [Then] abort without writing', async () => {
      const cwd = createMinimalProject({
        git: { dirty: true, init: true },
        fixtures: { configFile: false, templateFiles: false, base: true },
      });

      return Effect.gen(function* ($) {
        const args = ReadonlyArray.make('init');
        const fiber = yield* $(Effect.fork(Cli.run(args)));

        yield* $(MockTerminal.inputKey('enter'));

        yield* $(Fiber.join(fiber));

        const lines = yield* $(MockConsole.getLines({ stripAnsi: true }));
        expect(lines).toMatchInlineSnapshot(`
            [
              "Init",
              " Git ",
              "Checking working tree clean",
              "Git working tree dirty - proceed with caution.
            Recommendation: commit all changes before proceeding.",
              "? Continue? › yes / no",
              "✔ Continue? … yes / no
            ",
              "",
            ]
          `);

        expect(() =>
          fs.readFileSync(`${cwd}/scribe.config.ts`),
        ).toThrowErrorMatchingInlineSnapshot(
          `[Error: ENOENT: no such file or directory, open '${cwd}/scribe.config.ts']`,
        );
      }).pipe(runEffect(cwd));
    });
  });
});

it('[Given] --help flag [Then] print help information', async () => {
  const cwd = createMinimalProject();

  return Effect.gen(function* ($) {
    const args = ReadonlyArray.make('init', '--help');
    const fiber = yield* $(Effect.fork(Cli.run(args)));

    yield* $(Fiber.join(fiber));

    const lines = yield* $(MockConsole.getLines({ stripAnsi: true }));
    expect(lines).toMatchInlineSnapshot(`
        [
          "Scribe

        Scribe ${packageJson.version}

        USAGE

        $ init

        OPTIONS

        --completions sh | bash | fish | zsh

          One of the following: sh, bash, fish, zsh

          Generate a completion script for a specific shell

          This setting is optional.

        (-h, --help)

          A true or false value.

          Show the help documentation for a command

          This setting is optional.

        --wizard

          A true or false value.

          Start wizard mode for a command

          This setting is optional.

        --version

          A true or false value.

          Show the version of the application

          This setting is optional.
        ",
        ]
      `);
  }).pipe(runEffect(cwd));
});
