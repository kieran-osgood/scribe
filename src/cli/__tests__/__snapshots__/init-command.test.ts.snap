// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Given] Git Clean > [When] filepath full [then] print failure 2`] = `
"import type { ScribeConfig } from '@scribe/config';

const config = {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['./src/common/test-fixtures'],
  },
  templates: {
    component: {
      outputs: [
        {
          templateFileKey: 'component',
          output: {
            directory: 'examples/src/components',
            fileName: '{{Name}}.ts',
          },
        },
      ],
    },
    screen: {
      outputs: [
        {
          templateFileKey: 'screen',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.ts',
          },
        },
        {
          templateFileKey: 'screen.test',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.test.ts',
          },
        },
      ],
    },
  },
} satisfies ScribeConfig;

export default config;
"
`;

exports[`[Given] Git Clean > [When] user accepts [Then] create file 2`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;
